digdag-core/src/main/java/io/digdag/core/agent/ExtractArchiveWorkspaceManager.java

```diff
@@ -22,19 +31,53 @@
     public ExtractArchiveWorkspaceManager(TempFileManager tempFiles)
     {
         this.tempFiles = tempFiles;
+
     }
 
     @Override
     public <T> T withExtractedArchive(TaskRequest request, ArchiveProvider archiveProvider, WithWorkspaceAction<T> func)
             throws IOException
     {
-        try (TempDir workspacePath = createNewWorkspace(request)) {
-            Optional<StorageObject> in = archiveProvider.open();
-            if (in.isPresent()) {
-                ProjectArchives.extractTarArchive(workspacePath.get(), in.get().getContentInputStream());
+        TempDir workspacePath = null;
+        try {
+            try {
+                // A temporary workspace should be created for each try
+                // And always close the temporary workspace
+                workspacePath = RetryExecutor.retryExecutor()
+                        .retryIf(exception -> true)
+                        .withInitialRetryWait(EXTRACT_MIN_RETRY_WAIT_MS)
+                        .withMaxRetryWait(EXTRACT_MAX_RETRY_WAIT_MS)
+                        .onRetry((exception, retryCount, retryLimit, retryWait) ->
+                                logger.warn("Failed to extract archive: retry {} of {}", retryCount, retryLimit, exception))
+                        .withRetryLimit(EXTRACT_RETRIES)
+                        .run(() -> {
+                            TempDir newWorkSpacePath = null;
+                            try {
+                                newWorkSpacePath = createNewWorkspace(request);
+                                Optional<StorageObject> in = archiveProvider.open();
+                                if (in.isPresent()) {
+                                    ProjectArchives.extractTarArchive(newWorkSpacePath.get(), in.get().getContentInputStream());
+                                }
+                                return newWorkSpacePath;
+                            }
+                            catch (Throwable e) {
+                                if (newWorkSpacePath != null) {
+                                    newWorkSpacePath.close();
+                                }
+                                throw e;
+                            }
+                        });
+            }
+            catch (RetryExecutor.RetryGiveupException e) {
+                Throwables.propagate(e);
             }
             return func.run(workspacePath.get());
         }
+        finally {
+            if (workspacePath != null) {
```

Comment:
I think that this null check is unnecessary if you don't assign null when you declare `workspacePath`.
Or  how about using try-with-resource here like this?:

```java
try (TempDir  workspacePath = RetryExecutor.retryExecutor()...) {
    ...
}
catch (RetryExecutor.RetryGiveupException e) {
    ...
}
```

Reply:
One thing I wanted to do in this change is to create a new workspace for each try. So workspaces need to be created in `RetryExecutor.run()`.

Reply:
oh right. I was missing about it